File: composeApp/src/androidMain/kotlin/io/github/a7asoft/ui/ReaderScreen.kt
package io.github.a7asoft.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import io.github.a7asoft.ui.components.CardSummary
import io.github.a7asoft.ui.components.ErrorPanel
import io.github.a7asoft.ui.components.ReadingProgress
import io.github.a7asoft.ui.components.StatusHeader

/**
 * Pure-presentation screen for the reader flow.
 *
 * State and callbacks are hoisted; the screen renders only what the
 * caller passes. Test by passing a fake [ReaderUiState] and asserting
 * which sub-composable is shown.
 */
@Composable
public fun ReaderScreen(
    state: ReaderUiState,
    onTryAgain: () -> Unit,
    onOpenNfcSettings: () -> Unit,
) {
    Surface(modifier = Modifier.fillMaxSize()) {
        Column(
            modifier = Modifier.padding(24.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
        ) {
            StatusHeader(state, onOpenNfcSettings)
            ReaderContent(state, onTryAgain)
        }
    }
}

@Composable
@Suppress("CyclomaticComplexMethod")
// why: exhaustive `when` over the sealed `ReaderUiState` catalogue. Each
// branch is a single mapping, not an arm of real cyclomatic complexity.
private fun ReaderContent(state: ReaderUiState, onTryAgain: () -> Unit) {
    when (state) {
        ReaderUiState.NfcUnavailable, ReaderUiState.NfcDisabled, ReaderUiState.Idle -> Unit
        ReaderUiState.TagDetected -> ReadingProgress(stageLabel = "Tag detected")
        ReaderUiState.SelectingPpse -> ReadingProgress(stageLabel = "Selecting application directory")
        is ReaderUiState.SelectingApplication -> ReadingProgress(stageLabel = "Selecting application ${state.aid}")
        ReaderUiState.ReadingRecords -> ReadingProgress(stageLabel = "Reading records")
        is ReaderUiState.Done -> CardSummary(card = state.card, onTryAgain = onTryAgain)
        is ReaderUiState.Failed -> ErrorPanel(error = state.error, onTryAgain = onTryAgain)
    }
}

File: iosApp/iosApp/ui/ReaderScreen.swift
import SwiftUI

/// Top-level SwiftUI screen for the reader sample.
///
/// Pure presentation: state and callbacks come from the supplied
/// ``ReaderViewModel``. The screen renders the ``StatusHeader`` plus
/// one of ``ReadingProgress`` / ``CardSummary`` / ``ErrorPanel`` /
/// the Scan button depending on ``ReaderUiState``.
public struct ReaderScreen: View {

    @ObservedObject private var viewModel: ReaderViewModel

    /// Build a screen bound to the supplied view-model.
    public init(viewModel: ReaderViewModel) {
        self.viewModel = viewModel
    }

    public var body: some View {
        ScrollView {
            VStack(spacing: 24) {
                StatusHeader(state: viewModel.uiState)
                content
            }
            .padding(24)
        }
        .onAppear { viewModel.refreshAvailability() }
    }

    @ViewBuilder
    private var content: some View {
        switch viewModel.uiState {
        case .nfcUnavailable:
            EmptyView()
        case .idle:
            scanButton
        case .scanning:
            ReadingProgress(stageLabel: "Starting scan")
        case .tagDetected:
            ReadingProgress(stageLabel: "Tag detected")
        case .selectingPpse:
            ReadingProgress(stageLabel: "Selecting application directory")
        case .selectingApplication(let aidHex):
            ReadingProgress(stageLabel: "Selecting application \(aidHex)")
        case .readingRecords:
            ReadingProgress(stageLabel: "Reading records")
        case .done(let card):
            CardSummary(card: card, onTryAgain: viewModel.reset)
        case .failed(_, let friendly, let diagnostic):
            ErrorPanel(
                friendlyMessage: friendly,
                diagnostic: diagnostic,
                onTryAgain: viewModel.reset
            )
        }
    }

    private var scanButton: some View {
        Button(action: viewModel.startScan) {
            Text("Scan card")
                .frame(maxWidth: .infinity)
        }
        .buttonStyle(.borderedProminent)
        .controlSize(.large)
    }
}
